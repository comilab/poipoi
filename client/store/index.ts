import { defineComponent, provide, inject } from 'nuxt-composition-api'
import { InjectionKey } from '@vue/composition-api'

import sessionStore from '~/store/session'
import settingStore from '~/store/setting'
import toastStore from '~/store/toast'

type Store = {
  session: ReturnType<typeof sessionStore>
  setting: ReturnType<typeof settingStore>
  toast: ReturnType<typeof toastStore>
}

const makeStore = () => {
  return {
    session: sessionStore(),
    setting: settingStore(),
    toast: toastStore()
  }
}

export const key: InjectionKey<Store> = Symbol('Store')

export const store: Store = makeStore()

export const StoreProvider = defineComponent({
  setup () {
    provide(key, store)
  },
  render (h) {
    return h('div', this.$slots.default)
  }
})

export const useStore = () => inject(key)!

export const resetStore = () => {
  Object.assign(store, makeStore())
}

export const actions = {
  async nuxtClientInit () {
    try {
      await Promise.all([
        store.session.load(),
        store.setting.load()
      ])
    } catch (error) {
      //
    }
  }
}
