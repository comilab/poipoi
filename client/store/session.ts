import { reactive, toRefs, computed } from 'nuxt-composition-api'
import axios from 'axios'

import User from '~/models/User'

export default function sessionStore () {
  const state = reactive({
    user: null as User|null
  })

  const isLogin = computed(() => !!state.user)

  async function create (query: any) {
    await axios.get('/sanctum/csrf-cookie', {
      baseURL: process.env.apiUrl
    })
    await axios.post('/sessions', {
      ...query,
      device_name: 'poipoi-web'
    })
  }

  async function load () {
    try {
      const { data } = await axios.get('/sessions')
      setUser(new User(data))
    } catch (error) {
      setUser(null)
    }
  }

  async function destroy () {
    try {
      await axios.delete('/sessions/poipoi-web')
    } catch (error) {
      //
    }
    setUser(null)
  }

  function setUser (user: User|null) {
    state.user = user
  }

  return {
    ...toRefs(state),
    isLogin,
    create,
    load,
    destroy,
    setUser
  }
}
