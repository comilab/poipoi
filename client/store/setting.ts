import { reactive, toRefs } from 'nuxt-composition-api'
import axios from 'axios'

import useApi from '~/composables/use-api'
import Setting from '~/models/Setting'

export default function sessionStore () {
  const api = useApi('/settings', Setting)

  const state = reactive({
    data: null as Setting|null
  })

  async function load () {
    const { data } = await axios.get('/settings')
    setData(new Setting(data))
  }

  async function save (data: any) {
    await api.save(data)
    setData(api.item.value!)
  }

  function setData (data: Setting|null) {
    state.data = data
  }

  return {
    ...toRefs(state),
    load,
    save,
    setData
  }
}
