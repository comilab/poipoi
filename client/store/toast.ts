import { reactive, toRefs } from 'nuxt-composition-api'

export type Toast = {
  type: 'success'|'info'|'warning'|'danger'
  message: string
  timeout?: number
}

export default function toastStore () {
  const state = reactive({
    toasts: [] as Toast[]
  })

  function push ({ type, message, timeout = 5000 }: Toast) {
    const toast = {
      type,
      message,
      timeout
    }
    state.toasts.push(toast)
    if (timeout) {
      setTimeout(() => {
        remove(toast)
      }, timeout)
    }
  }

  function remove (toast: Toast) {
    const index = state.toasts.findIndex(t => t === toast)
    if (index >= 0) {
      state.toasts.splice(index, 1)
    }
  }

  return {
    ...toRefs(state),
    push,
    remove
  }
}
