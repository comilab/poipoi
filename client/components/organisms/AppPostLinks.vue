<template lang="pug">
ul.flex.justify-end
  li.mx-1(v-if="post.actualEnableTwitterShare")
    AppTwitterShareButton(
      :url="post.url",
      :text="post.caption",
      append-site-title,
      size="sm",
      breakpoint="md"
    )
  li.mx-1(v-if="store.session.isLogin")
    AppButton(
      :to="`${post.path}/edit`",
      :icon="faEdit",
      color="gray",
      size="sm",
      breakpoint="md"
    ) 編集
  li.mx-1(v-if="store.session.isLogin")
    AppButton(
      :icon="faTrashAlt",
      color="gray",
      size="sm",
      breakpoint="md",
      @click="showDeleteModal = true"
    ) 削除
  AppModal(
    v-model="showDeleteModal",
    :persistent="destroying"
  )
    template(#default="{ close }")
      .text-center.my-8 この投稿を削除してよろしいですか？
      .flex.justify-center
        AppButton.mx-2(
          color="blue",
          :icon="faTrashAlt",
          :indeterminate="destroying",
          :disabled="destroying",
          @click="deletePost"
        ) 削除する
        AppButton.mx-2(
          color="gray",
          :disabled="destroying",
          @click="close"
        ) キャンセル
</template>

<script lang="ts">
import { defineComponent, reactive, toRefs } from 'nuxt-composition-api'
import { faEdit, faTrashAlt } from '@fortawesome/free-solid-svg-icons'

import Post from '~/models/Post'
import { useStore } from '~/store'
import usePostsApi from '~/composables/use-posts-api'
import AppButton from '~/components/atoms/AppButton.vue'
import AppTwitterShareButton from '~/components/molecules/AppTwitterShareButton.vue'
import AppModal from '~/components/molecules/AppModal.vue'

export default defineComponent({
  components: {
    AppButton,
    AppTwitterShareButton,
    AppModal
  },
  props: {
    post: {
      type: Post,
      required: true
    }
  },
  setup (props, { emit }) {
    const store = useStore()

    const { destroy, destroying } = usePostsApi(props.post)

    const localState = reactive({
      showDeleteModal: false
    })

    const deletePost = async () => {
      await destroy()
      emit('deleted')
    }

    return {
      store,
      destroying,
      ...toRefs(localState),
      deletePost,
      faEdit,
      faTrashAlt
    }
  }
})
</script>
