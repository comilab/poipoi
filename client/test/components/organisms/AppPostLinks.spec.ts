import { fireEvent, render, waitFor } from '@testing-library/vue'
import { mocked } from 'ts-jest/utils'
import { ref } from 'nuxt-composition-api'

import AppPostLinks from '~/components/organisms/AppPostLinks.vue'
import * as store from '~/store'
import settingFactory from '~/test/factories/setting'
import postFactory from '~/test/factories/post'
import userFactory from '~/test/factories/user'
import usePostsApi from '~/composables/use-posts-api'

jest.mock('~/composables/use-posts-api')

describe('components/organisms/AppPostLinks', () => {
  let options: any
  let postsApiMock: any

  beforeEach(() => {
    options = {
      props: {
        post: postFactory.build()
      },
      stubs: {
        FontAwesomeIcon: true,
        NuxtLink: true
      }
    }
    store.store.setting.setData(settingFactory.build())
    jest.spyOn(store, 'useStore').mockReturnValue(store.store)

    postsApiMock = {
      destroy: jest.fn().mockResolvedValue(null),
      destroying: ref(false)
    }
    mocked(usePostsApi).mockReturnValue({ ...postsApiMock })
  })

  test('renders correctly', () => {
    store.store.session.setUser(userFactory.build())
    store.store.setting.setData(settingFactory.build({
      siteTitle: 'site-title'
    }))
    options.props.post = postFactory.build({
      url: 'https://example.com',
      caption: 'caption',
      actualEnableTwitterShare: true
    })
    const { container } = render(AppPostLinks, options)

    expect(container).toMatchSnapshot()
  })

  describe('AppTwitterShareButton', () => {
    test('actualEnableTwitterShareがtrueなら表示', async () => {
      options.props.post = postFactory.build({
        actualEnableTwitterShare: false
      })
      const { queryByRole, updateProps } = render(AppPostLinks, options)

      expect(queryByRole('link', { name: 'Tweet' })).not.toBeInTheDocument()

      await updateProps({
        post: postFactory.build({
          actualEnableTwitterShare: true
        })
      })
      expect(queryByRole('link', { name: 'Tweet' })).toBeInTheDocument()
    })
  })

  describe('編集', () => {
    test('ログイン中のみ表示', async () => {
      store.store.session.setUser(null)
      const { queryByText } = render(AppPostLinks, options)
      expect(queryByText('編集')).not.toBeInTheDocument()

      store.store.session.setUser(userFactory.build())
      await waitFor(() => {
        expect(queryByText('編集')).toBeInTheDocument()
      })
    })
  })

  describe('削除', () => {
    beforeEach(() => {
      store.store.session.setUser(userFactory.build())
    })

    test('ログイン中のみ表示', async () => {
      const { queryByRole } = render(AppPostLinks, options)
      expect(queryByRole('button', { name: '削除' })).toBeInTheDocument()

      store.store.session.setUser(null)
      await waitFor(() => {
        expect(queryByRole('button', { name: '削除' })).not.toBeInTheDocument()
      })
    })

    test('クリックするとモーダルを表示', async () => {
      const { getByRole, getByText } = render(AppPostLinks, options)

      await fireEvent.click(getByRole('button', { name: '削除' }))

      expect(getByText('この投稿を削除してよろしいですか？')).toBeInTheDocument()
    })

    test('モーダル内でキャンセルを押すと閉じる', async () => {
      const { getByRole, queryByText } = render(AppPostLinks, options)

      await fireEvent.click(getByRole('button', { name: '削除' }))
      await fireEvent.click(getByRole('button', { name: 'キャンセル' }))

      expect(queryByText('この投稿を削除してよろしいですか？')).not.toBeInTheDocument()
    })

    test('モーダル内で削除するを押すと削除処理を実行', async () => {
      const { getByRole } = render(AppPostLinks, options)

      await fireEvent.click(getByRole('button', { name: '削除' }))
      await fireEvent.click(getByRole('button', { name: '削除する' }))

      expect(postsApiMock.destroy).toBeCalled()
    })

    test('削除中はモーダル内ボタンを押せない', async () => {
      const { getByRole } = render(AppPostLinks, options)

      await fireEvent.click(getByRole('button', { name: '削除' }))

      expect(getByRole('button', { name: '削除する' })).not.toHaveAttribute('disabled')
      expect(getByRole('button', { name: 'キャンセル' })).not.toHaveAttribute('disabled')

      postsApiMock.destroying.value = true

      await waitFor(() => {
        expect(getByRole('button', { name: '削除する' })).toHaveAttribute('disabled')
        expect(getByRole('button', { name: 'キャンセル' })).toHaveAttribute('disabled')
      })
    })

    test('削除中はモーダルを閉じられない', async () => {
      const { queryByLabelText, getByRole } = render(AppPostLinks, options)

      await fireEvent.click(getByRole('button', { name: '削除' }))

      expect(queryByLabelText('close')).toBeInTheDocument()
      expect(getByRole('button', { name: 'キャンセル' })).not.toHaveAttribute('disabled')

      postsApiMock.destroying.value = true

      await waitFor(() => {
        expect(queryByLabelText('close')).not.toBeInTheDocument()
      })
    })

    test('処理終了後はdeletedイベントを発火', async () => {
      const { getByRole, emitted } = render(AppPostLinks, options)

      await fireEvent.click(getByRole('button', { name: '削除' }))
      await fireEvent.click(getByRole('button', { name: '削除する' }))

      await waitFor(() => {
        expect(emitted().deleted).toHaveLength(1)
      })
    })
  })
})
