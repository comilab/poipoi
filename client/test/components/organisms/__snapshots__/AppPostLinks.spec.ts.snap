// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/AppPostLinks renders correctly 1`] = `
<div>
  <ul
    class="flex justify-end"
  >
    <li
      class="mx-1"
    >
      <a
        breakpoint="md"
        class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none tweet-button bg-blue-600 text-white hover_bg-blue-700 text-sm px-2 py-1 square-sm breakpoint-md rounded"
        data-testid="button"
        href="http://twitter.com/share?url=https%3A%2F%2Fexample.com&text=caption+-+site-title"
        size="sm"
        target="_blank"
      >
        <div>
          <fontawesomeicon-stub
            class="fa-fw md_mr-2"
            data-testid="icon"
            icon="[object Object]"
          />
        </div>
        <div
          class="hidden md_block"
          data-testid="text"
        >
          Tweet
        </div>
      </a>
    </li>
    <li
      class="mx-1"
    >
      <nuxtlink-stub
        class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-gray-300 text-gray-800 hover_bg-gray-400 text-sm px-2 py-1 square-sm breakpoint-md rounded"
        data-testid="button"
        to="/posts/2/edit"
      >
        <div>
          <fontawesomeicon-stub
            class="fa-fw md_mr-2"
            data-testid="icon"
            icon="[object Object]"
          />
        </div>
        <div
          class="hidden md_block"
          data-testid="text"
        >
          編集
        </div>
      </nuxtlink-stub>
    </li>
    <li
      class="mx-1"
    >
      <button
        class="inline-flex items-center justify-center whitespace-no-wrap transition-colors duration-300 disabled_opacity-75 disabled_pointer-events-none focus_outline-none bg-gray-300 text-gray-800 hover_bg-gray-400 text-sm px-2 py-1 square-sm breakpoint-md rounded"
        data-testid="button"
        type="button"
      >
        <div>
          <fontawesomeicon-stub
            class="fa-fw md_mr-2"
            data-testid="icon"
            icon="[object Object]"
          />
        </div>
        <div
          class="hidden md_block"
          data-testid="text"
        >
          削除
        </div>
      </button>
    </li>
    <transition-stub
      enterclass="opacity-0"
      leavetoclass="opacity-0"
    >
      <!---->
    </transition-stub>
  </ul>
</div>
`;
